Show unsigned Firefox builds if necessary

danielhjacobs opened the pull request:
~~Currently, this will show the latest available Firefox build (either signed or unsigned) in the latest available column instead of both or only the signed one. I want to fix that, but I don't want to show a strike-through for the other rows, so release-assets would need some conditional logic that we may want to avoid.~~

danielhjacobs commented:
Before I forget, the logic I want to add is:

- nightly shows signed if exists, otherwise unsigned if exists, otherwise strike-through
- "latest" always shows signed and shows unsigned if unsigned date is greater than signed date (finding a way to check this may be difficult) 
- Show the unsigned instructions if anything unsigned is present

I won't attempt to write any of this logic until tonight's assets are released since that will give me something to work with.

danielhjacobs commented:
For bullet three, an idea to avoid an extra Jekyll would be by utilizing JavaScript, spans, and a div. Put the unsigned FF instructions in `<div id="unsigned_ff_instructions">...</div>`. Surround the unsigned releases with `<span name="unsigned">...</span>`. Add to https://github.com/ruffle-rs/ruffle-rs.github.io/blob/master/_includes/footer.html#L10:

    if ($("span[name='unsigned']").length) {
        $('unsigned_ff_instructions').show()
    }

danielhjacobs commented:
New logic added as described in previous comments. If there are no unsigned builds in any table rows, the website will look exactly as it does right now.

If there are unsigned builds in the table, the Usage section will look like this (zoomed out to show everything on one screen):
![ruffle_website_with_unsigned_build](https://user-images.githubusercontent.com/29206584/134271597-29578688-c244-4122-8d66-b6784d70decc.png)

And the downloads section will look like this (zoomed out to show everything on one screen):
![ruffle_downloads_with_unsigned_build](https://user-images.githubusercontent.com/29206584/134271671-188472a0-2ae9-4343-b726-a79f202c337a.png)

midgleyc commented:
Very neat!

I think:
* both signed and unsigned instructions should always show, because I personally prefer that webpages don't dynamically change  distant sections based on the content of other sections -- if I found the instructions there at one point, I'd like them to always be there
* the signed Firefox instructions should be above the unsigned, because I hope that's the more standard case
* instead of "not unsigned", how about "standard" or something?

danielhjacobs commented:
I am a bit torn on the first bullet in your feedback, but I implemented it anyway (last and in its own commit so it's easy to revert). On one hand, I tend to agree with wanting webpages to not dynamically change their wording. On the other hand, if and when this issue is fixed, I expect there to not be unsigned builds in the table for at least a few months, so instructions about something that does not appear on the website may be confusing to end users.

I also implemented a slight styling change; I felt there was too much white-space above section headers, so I reduced it from 10rem to 5rem.

danielhjacobs commented:
I believe this PR matches midgleyc's suggestions now. I'd appreciate a review from @relrelb. The Firefox extension on the website is now over 2 and a half weeks out of date, and the signing issue shows no sign of fixing itself.

relrelb commented on index.html:
```diff
@@ -147,13 +147,24 @@ <h4>Chrome</h4>
-						<h4>Firefox</h4>
+					</div>
+					<div class="col-base col-lg col-lg-4">
+						<h4>Firefox (standard)</h4>
```
Maybe this should be called "Firefox (signed)"?

danielhjacobs replied:
Unless I also changed the name in the table to (signed), I'm not sure end-user would know what signed meant, especially if one day the signing issue fixes itself and there are no unsigned builds in the table.

relrelb commented on _includes/release-assets.html:
```diff
@@ -31,13 +31,31 @@
 		{% else %}
 			<li style="text-decoration: line-through">Firefox</li>
 		{% endif %}
+		{% if unsigned_firefox != nil and unsigned_firefox.name > signed_firefox.name %}
```
Also, I'm not sure it that's possible, but what if `unsigned_firefox != nil` and `signed_firefox == nil`? Then I assume you'll get an error due to `nil`-deref.

relrelb commented on _includes/release-assets.html:
```diff
@@ -31,13 +31,31 @@
-		{% if firefox != nil %}
-			<li><a href="{{ firefox.browser_download_url}}" target="_blank">Firefox</a></li>
+		{% if include.latest %}
+		
```
Please remove trailing spaces.

relrelb commented on _includes/release-assets.html:
```diff
@@ -31,13 +31,31 @@
-		{% assign firefox = include.assets | where_exp:"item", "item.name contains '-firefox.'" | first %}
-		{% if firefox != nil %}
-			<li><a href="{{ firefox.browser_download_url}}" target="_blank">Firefox</a></li>
+		{% if include.latest %}
```
I've managed to compact the logic a bit to avoid repeating code:
```
{% if signed_firefox != nil %}
	<li><a href="{{ signed_firefox.browser_download_url}}" target="_blank">Firefox</a></li>
{% elsif !include.latest && unsigned_firefox != nil %}
	<li><a href="{{ unsigned_firefox.browser_download_url}}" target="_blank">Firefox (unsigned)</a></li>
{% else %}
	<li style="text-decoration: line-through">Firefox</li>
{% endif %}

{% if include.latest && (signed_firefox == nil || unsigned_firefox != nil and unsigned_firefox.created_at > signed_firefox.created_at) %}
	<li><a href="{{ unsigned_firefox.browser_download_url}}" target="_blank">Firefox (unsigned) - More recent</a></li>
{% endif %}
```
Note that I never used Jekyll and I haven't tested it so some syntax might be wrong. Also my boolean logic might be flawed...

danielhjacobs replied:
Mostly correct, slightly changed to work correctly (and not &&, or not ||, and !include.latest is not a proper test for nil in Jekyll, so include.latest == nil instead).

relrelb commented on assets/title.scss:
```diff
@@ -120,7 +120,7 @@ table.table-dark th {
 
 .section h2 {
-  padding-top:10rem;
+  padding-top:5rem;
```
I haven't setup Jekyll so I didn't test that properly (I edited the styles manually in DevTools), and I trust you the style changes are fine.

relrelb commented on _includes/release-assets.html:
```diff
@@ -31,12 +31,19 @@
-			<li><a href="{{ firefox.browser_download_url}}" target="_blank">Firefox</a></li>
+		{% if signed_firefox != nil %}
+		<li><a href="{{ signed_firefox.browser_download_url}}" target="_blank">Firefox</a></li>
+		{% elsif include.latest == nil and unsigned_firefox != nil %}
```
Can't `include.latest == nil` be `not include.latest`? Quite hypothetic, but what if someone passes `latest = false`?

danielhjacobs replied:
https://github.com/Shopify/liquid/issues/138#issuecomment-657514982

danielhjacobs replied:
I could just pass latest=false explicitly, but then there's the opposite problem of that always having to be done. Your choice.

relrelb replied:
Maybe it could be `include.latest != true`?

relrelb reviewed (approved):
LGTM, Thank you!
Let's see how it works...
